Add boxsets ability

the-black-eagle opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

DaveTBlake committed:
Multi-discs albums navigate albums > discs > songs
- add discs node
- Remove unneeded new DirectoryNode descendants
- Handle "* All Albums", "* All Discs"

Fixups to compile

Hack with *all and using xml name for title of folder type nodes (like filter type does)

the-black-eagle committed:
Add boxset detection and navigation

the-black-eagle committed:
Add disc title to Estuary

the-black-eagle committed:
Split on disc title if present, else split on disc number

Localize 'Disc' when generating title for forced boxsets

DaveTBlake commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -12764,7 +12767,7 @@ msgctxt "#20340"
 msgstr ""
 
-#empty string with id 20341
+#empty string with id #20341
```
Unnecessary change

DaveTBlake commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -21145,8 +21148,24 @@ msgctxt "#38073"
 
-#empty strings from id 38074 to 38099
-#strings 38074 to 38099 reserved for music library
+#: system/library/music/musicroles/boxsets.xml
```
My error system/library/music/boxsets.xml

DaveTBlake commented on xbmc/filesystem/LibraryDirectory.cpp:
```diff
@@ -65,12 +65,16 @@ bool CLibraryDirectory::GetDirectory(const CURL& url, CFileItemList &items)
       {
+        std::string label;
+        if (XMLUtils::GetString(node, "label", label))
+          label = CGUIControlFactory::FilterLabel(label);
```
All changes to CLibraryDirectory can be dropped, they are in PR 16626

DaveTBlake commented on xbmc/filesystem/MusicDatabaseDirectory/DirectoryNodeAlbum.cpp:
```diff
@@ -22,13 +22,28 @@ CDirectoryNodeAlbum::CDirectoryNodeAlbum(const std::string& strName, CDirectoryN
+    return NODE_TYPE_SONG;
+
+  //! @todo: This makes all multi-disc albums album > disc > song, make disc level optional
+  CMusicDatabase musicdatabase;
```
Add a new music lib setting to enable(default)/disable navigation from multi-disc albums to discs rather than directly to songs, and use it *here* to allow albums > discs navigation to be optional. That way we can give users original behaviour if they want it.

A search for `musiclibrary.showcompilationartists` may help you find example of how to do a new setting

the-black-eagle replied:
OK, I'll give it a go.

the-black-eagle replied:
Done that.  Default is false and navigation is as it was.  Toggling it on enables disc type navigation.  Incidentally, extraart still works in original navigation if you have set any up, so the covers change as you scroll up and down the tracks between the discs.

DaveTBlake replied:
Users won't see the new grouping as discs feature that way around. Make default having a "discs" level, and they can change the setting to get old behaviour back

the-black-eagle replied:
Done.  Not 100% sure that it should be turned off for boxsets at all, given that's where we started with all this, but as I can't see a particularly clean way of doing that I'll go with it.

DaveTBlake replied:
Humm, yes when viewing boxed sets you do always want album > disc...
If I think of a way to do that nicely I'll let you know. At least on by default means users with boxed sets will see the immediate advantage.

DaveTBlake commented on xbmc/filesystem/MusicDatabaseDirectory/DirectoryNodeBoxsetDiscs.cpp:
```diff
@@ -0,0 +1,57 @@
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DirectoryNodeBoxsetDiscs.h"
```
`CDirectoryNodeBoxsetDiscs` needs renaming  `CDirectoryNodeDiscs` throughout because it applies to navigation generally not just boxed stes

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -1893,6 +1910,35 @@ bool CMusicDatabase::GetSongsByArtist(int idArtist, std::vector<int> &songs)
 };
 
+bool CMusicDatabase::GetSongByDiscSubtitleAndAlbum(const std::string& strDiscSubtitle, int idAlbum)
+{
```
Unused funtion - remove.

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -6328,7 +6378,7 @@ bool CMusicDatabase::GetSongsByWhereJSON(const std::set<std::string>& fields, co
         joinLayout.SetField(joinToSongs_idAlbumArtist, 
           JSONtoDBSong[index_idAlbumArtist + joinToSongs_idAlbumArtist].SQL);
-      }      
+      }     
```
Cosmetic whitespace change

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -7924,7 +7980,7 @@ int CMusicDatabase::UpdateSource(const std::string& strOldName, const std::strin
     if (!strOldName.empty())
     {
-      strSQL = PrepareSQL("SELECT idSource, strMultipath FROM source WHERE strName LIKE '%s'", 
+      strSQL = PrepareSQL("SELECT idSource, strMultipath FROM source WHERE strName LIKE '%s'",
```
Cosmetic whitespace change

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -8203,7 +8259,7 @@ bool CMusicDatabase::MigrateSources()
     // Empty sources table (related link tables removed by trigger);
     ExecuteQuery("DELETE FROM source");
-  
+
```
Cosmetic whitespace change

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -8321,6 +8377,7 @@ bool CMusicDatabase::GetSourcesByArtist(int idArtist, CFileItem* item)
         m_pDS->close();
         return true;
       }
+
```
Cosmetic whitespace change

the-black-eagle replied:
Grrr, thought I had fixed 'em all !!

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -8888,6 +8951,100 @@ bool CMusicDatabase::GetGenresJSON(CFileItemList& items, bool bSources)
+  // adding disc number and subtitle per item
+  // ! @TODO: Convert into GetDiscsByWhere,  handle no idAlbum => all discs
+  // 1 disc => same as album, is acceptable. Maybe use JOIN rather than 2 queries?
+
```
`GetBoxsetDiscs` should be renamed `GetDiscsByWhere` as used for any multi-disc album not just boxed sets. 
Also consider effect of other filter/options e.g. say showing just albums that have songs of genre = X (and nothing says all the songs on an album have the same genre), or have an artist filter and showing those albums with songs by an artists including guest appearences and on compilations. Navigation to discs should just show those discs with songs that staify the criteria. Look at how GetAlbumsByWhere uses GetFilter to do that for hints (or ask)

the-black-eagle replied:
Do you mean, if an album has four discs and three discs contain tracks of genre 'pop' and one disc tracks of 'rock', you want to just display either the three discs or the one, depending upon the filter ?? And similar for an artist or any other filter criteria ?

DaveTBlake replied:
Sorry for leaving you hanging, my limited Kodi time has been eaten by other things.
Yes, to make the navigation similar to other nodes (but may need some consideration what behaviour is wanted). Artists is a better example: say you pick an artist from "artists" node, albums are listsed including an album that is a boxed set where the artist makes a guest appearance on 2 songs on different discs. When you click on that album what is displayed - discs or songs? You could say just a few song go directly to them, but we don't skip the album level when the artist only has one album so should we skip discs?

the-black-eagle replied:
We all have lives and limited time, so no worries on that score. I won't have much time myself over the next three days or so....
Anyway,  I have thought about this quite a lot.  I'm not convinced myself that it's actually needed, however that's probably for a different day.  In the meantime, from an implementation point of view, GetFilter() currently only understands a type of 'albums'  'artists' or 'songs'.  I have contemplated adding a 'discs' type (which might come in useful further along the line ??) but haven't done so yet because basically we are actually filtering the songs that are on a particular disc.  Discs themselves don't have artists or genres or roles and there is no structure to link them to such things.
With that in mind, I have been playing with making GetFilter always use a type of 'songs' if 'disc' is present as an option in the url.  This makes it include 'discid' and 'disctitle' (both of which are song options anyway) and then apply any other filter rules (for songs) in the same way they get applied for other routines that call it. 

If I don't do it like that, GetFilter sees a type of 'albums' but trying to filter discs by album doesn't make sense as you have to have an album first before you can get any discs for it.
I'm still not convinced that I want to be able to return say 15 individual discs that have an artist of 'blah' and a genre of 'foo'.  If I get back discs with no actual titles (the most common), I could end up with a list of 'disc 1, disc 1, disc 3, disc 2, disc 1, disc 1, etc' .

DaveTBlake replied:
OK, I'll see if I can give more clarity on what I had in mind here too. But that will be later over the weekend :)

DaveTBlake commented on xbmc/music/MusicDatabase.h:
```diff
@@ -202,6 +206,7 @@ class CMusicDatabase : public CDatabase
   bool SetSongUserrating(const std::string &filePath, int userrating);
   bool SetSongUserrating(int idSong, int userrating);
   bool SetSongVotes(const std::string &filePath, int votes);
+  bool GetSongByDiscSubtitleAndAlbum(const std::string& strDiscSubtitle, int idAlbum);
```
Unused function

DaveTBlake commented on xbmc/music/windows/GUIWindowMusicNav.cpp:
```diff
@@ -241,6 +241,8 @@ std::string CGUIWindowMusicNav::GetQuickpathName(const std::string& strPath) con
   else if (path == "special://musicplaylists/")
     return "Playlists";
+  else if (path == "musicdb://discs/")
+    return "discs";
```
CGUIWindowMusicNav::GetQuickpathName is deprecated function so can remove this change - see PR16626

DaveTBlake commented on xbmc/music/windows/MusicFileItemListModifier.cpp:
```diff
@@ -58,8 +58,8 @@ void CMusicFileItemListModifier::AddQueuingFolder(CFileItemList& items)
-  // No need for "all" when overview node and child node albums or artists
-  if (directoryNode->GetType() == NODE_TYPE_OVERVIEW &&
+  // No need for "all" when overview node and child node albums or artists without options
+  if (directoryNode->GetType() == NODE_TYPE_OVERVIEW && musicUrl.GetOptions().empty() &&
```
Again this area is changed is in PR16626, so no need here. Watch out when rebasing!

DaveTBlake reviewed (commented):
OK, decided to make some comments on this as it is. More to come I'm sure

DaveTBlake commented:
Darn, 3 more commits while I was posting. Oh well I'm sure we will manage to unravel it

DaveTBlake commented on xbmc/music/MusicDatabase.h:
```diff
@@ -128,6 +128,7 @@ class CMusicDatabase : public CDatabase
    \param iTrack [in] the track number and disc number of the song
    \param iDuration [in] the duration of the song
    \param iYear [in] the year of the song
+  \param strDiscSubtitle [in] subtitle of a disc if it belongs to a box-set
```
Not just boxed sets - "subtitle of a disc."

DaveTBlake commented on xbmc/music/MusicDatabase.h:
```diff
@@ -174,6 +176,7 @@ class CMusicDatabase : public CDatabase
    \param iTrack [in] the track number and disc number of the song
    \param iDuration [in] the duration of the song
    \param iYear [in] the year of the song
+   \param strDiscSubtitle [in] subtitle of a disc if it belongs to a box-set
```
Not just boxed sets - "subtitle of a disc."

DaveTBlake commented on xbmc/music/MusicThumbLoader.cpp:
```diff
@@ -298,6 +298,15 @@ bool CMusicThumbLoader::FillLibraryArt(CFileItem &item)
         if (artitem.artType == "thumb" && !bDiscSetThumbSet)
           item.SetArtFallback(artitem.artType, artname);
+        else if (StringUtils::StartsWith(artitem.artType, "cover"))
+        {
```
What are you doing hard coding an art type???? Anyway "Thumb" is the cover.
Need to look at more but this doen't seem right

the-black-eagle replied:
Yeah, perhaps my bad, but "thumb" is hard-coded.  Is fanart1, fanart2 etc not hardcoded then also ?

DaveTBlake replied:
Originally all art types were hard coded, v18 brought revolution that liberated art (meaning that skinners and addons could had a lot more freedom without needing future core changes) but had to leave some things for backwards compatibility. "Thumb" and "fanart" are still there, and there is more to do which will get rid of much of the separate treatment for "fanart". It is all historic. 

You have done pretty much what I did (and probably every dev before us) attempted to copy what is already there. Unfortunately some bits need not to be copied, but nothing is documented so you have no way to know until someone with more experience says "not like that".

I need to take a bit more time over it and see if I can point you in the right direction. It just came as a surprize give all the things in my head, but really no so surprize because there is no reason for you to know otherwise.

EDIT
So not "your bad" in any way, but hopefully my replacement art commit will make sense of how I was seeing this (with my knowledge of "art mysteries"

DaveTBlake commented on xbmc/music/Song.cpp:
```diff
@@ -221,6 +222,7 @@ void CSong::Serialize(CVariant& value) const
   value["duration"] = iDuration;
   value["track"] = iTrack;
   value["year"] = iYear;
+  //value['disctitle'] = strDiscSubtitle;
```
Commented out to get it to compile (on Windows) - investiate when JSON implemented

DaveTBlake commented on xbmc/music/infoscanner/MusicInfoScanner.cpp:
```diff
@@ -2124,8 +2180,9 @@ void CMusicInfoScanner::SetDiscSetArtwork(CAlbum& album, const std::vector<std::
-    return;  // No disc subfolders to process
+//  if (paths.size() <= 1)
+//    return;  // No disc subfolders to process
+//  Keep scanning - sets can be in one folder using extraart
```
Humm...... use of extraart approach (rather than per folder) is something to be considered generically for more than just box sets. I need to check deeper.

DaveTBlake commented on xbmc/music/tags/TagLoaderTagLib.cpp:
```diff
@@ -1048,10 +1057,27 @@ void CTagLoaderTagLib::SetGenre(CMusicInfoTag &tag, const std::vector<std::strin
+  {
     tag.SetMusicBrainzReleaseType(StringUtils::Join(values, CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_musicItemSeparator));
+    if (tag.GetMusicBrainzReleaseType().find("boxset") != std::string::npos)
+    {
```
Wondering if all this manipulation of  this tag value to get boxed set flag should happen when saving to the db (so in CMusicDatabase or CSong/CAlbum  rather than here in the tag reading process.

DaveTBlake reviewed (commented):
OK, looks like most of my previous commenst still hold, so here are a few more

DaveTBlake commented:
OK, moving discussion generally to this PR so things don't get lost.

I don't think you have done the art right, although we may need some more work skin side to do art in the modern approach. But I have not had time yet to get into it and it would be best for me to comment in one go than drip feed you hints.

The use of disc title for grouping discs isn't as bad as I theorised it would be (so much for holding all the design in my head). That is a thaw in my view, but not quite a win for you yet!  

Item "*all discs" does give you all the songs (yes, I know why not call it "*all songs", well historically it have always been "All [content name]"  not "all [child thing you are going to when clicked]) which is consistent with the other uses, but seeing "discs " and knowing that isn't really discs in the list beneath niggles me a little.

We  end up with paths like
musicdb://albums/1429/2/?albumid=1429&disctitle=SubtitleBoth
where the discnum in the first part of the url e.g. 2 is only the number of the first disc with the title given in the options e.g. "SubtitleBoth". We need to take care when that number used - getting the disc title, that is fine and (at the moment) for getting art which also OK, but need to check when doing other filtering etc. it doesn't bite us.

What about if the discs with matching titles are not adjacent in number e.g. disc 1 and 3 have the same title but disc 2 is different? Should grouping still apply?

DaveTBlake commented:
Another aspect entirely: what about the album infomation dialog?
At the moment bringing this up from the list of discs/disc groups gives the album information. Is that good, or do you want a disc specific variant? 
This info includes a list of songs (not all skins show it), should album info dialog give a list of discs instead?  Should a new disc info dialog show the list of songs? Or are these lists on the dialog a total waste of time since the normal node navigation soon shows lists of songs and with more details?

the-black-eagle committed:
Add settings toggle to enable/disable disc navigation in GUI

Rename 'DirectoryNodeBoxsetDiscs' to 'DirectoryNodeDiscs'

Make disc nav the default, rename GetBoxsetDiscs to GetDiscsByWHere

the-black-eagle commented:
> What about if the discs with matching titles are not adjacent in number e.g. disc 1 and 3 have the same title but disc 2 is different? Should grouping still apply?

Good question and one I hadn't considered!  If I say "no" here, then that messes up what I've done because it needs a lot more logic.  On the other hand, my gut reaction was indeed "no".  Tricky !!  Is it likely though that that would occur ?  Possibly, given the state of some users tags.  Have to think about that a bit more.

Album info dialog.  It shows what it should show I think. I don't know if you could show information about individual discs because I'm really not sure that there is any where to scrape that information from. Having checked 3 different sets on theaudiodb, last.fm and allmusic the only info available was for one set on last.fm which merely mentioned that it was a remaster of the original album (although there are 3 other discs with it).  Of course there are always nfo files that could add that information if there was somewhere to display it.  More info regarding a disc could be useful (remaster of original album, live version recorded at wherever, etc etc) but I don't think you need a list of discs because you get that when you open the album anyway.  Besides, its cumbersome navigating that dialog with a remote.  <info> - <up> - <up> - <right> - <right> to see the tracks.  Easier to press enter and have done with it.  Album description is on the left in any case. 

Art.  I may well have done it wrong.  I'd prefer that I didn't have a great long list of stuff to add in advancedsettings but being able to add extra art types that way seemed  reasonable so I just picked a type to use.

the-black-eagle commented:
Actually, having thought about it a little more, grouping won't apply if you try to group 1&3 together if disc 2 has a different title as it works on title == preceding title. All three should be split but 1&3 should still show the same title.  I'll test that at some point tomorrow.

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -9010,7 +9024,14 @@ bool CMusicDatabase::GetBoxsetDiscs(const std::string& strBaseDir, CFileItemList
+        {
+          if (coverart[i] == StringUtils::Format("cover%i", discnum))
+            pItem->SetArt("poster", coverart[i+1]); // set the individual cover art url
+        }
```
Disc specific cover art is already automatically fetched from folder.jpg held local to the disc music files and held as album art of type "thumb1", "thumb2", "thumb3" etc.  not "poster".

the-black-eagle replied:
Estuary is looking for a "poster" value in some of it's views.  This is why art shows in 2 of the views (but not the disc subtitle, just the original album title), not at all in the widelist view (just text, shows the disc title as text) but not the last.  If I set the 'poster' value, it does.  At the time, this seemed like a better (easier ?) option than faffing around with the skin although that could well have been the wrong approach when I think about it now.  Still, it worked !!  I tried various different approaches to try to get the custom art into 'thumb' without much success.  Similarly, there doesn't appear to be a way (not that I can see anyway) to alter or overwrite an album title so that it shows in all the different views.  It's not a big thing, and I can navigate alright with the info thats there but it isn't 100% correct in some views and that niggles me.  Guess it might need either another info label, some skin changes or a bit of both.

DaveTBlake replied:
Estuary is simply wrong (it was done by people that never use music lib and didn't give an dam). They even opened a forum thread for user bug feedback during beta while privately in team saying they had no intention of looking at any of it. And got praised! I have never forgiven them for such attitudes, (most) users deserve better. [Friday night honesty!]

I understand your desire and happiness at getting something that worked, and I'm sure what you did had reasons given what you could see, but yes some of it needs to be done in the skin. Also Confluence is much easier to follow and mod if you want to mock up something for testing.

the-black-eagle replied:
I can switch to using Confluence on my Dev system.  My actual 'production' systems use a heavily customised version of Estuary done by a forum member.  This version was never released publicly, but has some major modifications to the PVR section.  As we can only watch TV via Kodi & tvheadend the pvr part needs to be user friendly and this members mod is awesome.  I also rolled out a build of my branch to a libreelec install on one of my Pi's and to my main instance so that I can use the discs feature.  Both of these are using this modded skin (which I further modded to work with my radio helper) so that's why I looked at Estuary.  That and the fact that if something gives unexpected results in a skin, the first question is always "What happens if you use Estuary ?"

There are people that never use the music library !!??  Clearly they are missing out then.  It does go some way though to explaining why most views are set up with video library variables and some don't even contain any references to the music library, despite the same view being used for both.

DaveTBlake replied:
I always suggest users try a simpler skin like "Estuary _or Confluence_" ;)
You were right to test with Estuary, but wrong (understandably) to assume what it did was correct. Skinning isn't my thing so to test the art stuff I somethimes ask a team member to make me a skin hack, or I use debug a lot. If I can see that the art is there in the item artmap then that is good enough for me!

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -10683,12 +10704,14 @@ bool CMusicDatabase::GetArtForItem(int songId, int albumId, int artistId, bool b
       artitem.artType = m_pDS2->fv("type").get_asString();
+      if (discId > 0)
+        if (artitem.artType == StringUtils::Format("discart%i", discId))
+          artitem.artType = "discart";
```
Need to handle any type of disc specific art not just discart. OK for now that is the only thing available, but avoidng hard coded art types in core. 
But anyway by doing all the art handling in `CMusicThumbLoader::FillLibraryArt` can avoid adding discid parameter

DaveTBlake reviewed (commented):
OK, here is my suggestion for handling disc art
Drop your art commit (rebase with skip), and just add the following bits of code to  `CMusicThumbLoader::FillLibraryArt` instead
```
diff --git a/xbmc/music/MusicThumbLoader.cpp b/xbmc/music/MusicThumbLoader.cpp
index 909b614..a60f97b 100644
--- a/xbmc/music/MusicThumbLoader.cpp
+++ b/xbmc/music/MusicThumbLoader.cpp
@@ -270,6 +270,7 @@ bool CMusicThumbLoader::FillLibraryArt(CFileItem &item)
     std::string fanartfallback;
     bool bDiscSetThumbSet = false;
     std::map<std::string, std::string> artmap;
+    std::map<std::string, std::string> discartmap;
     for (auto artitem : art)
     {
       /* Add art to artmap, naming according to media type.
@@ -289,6 +290,21 @@ bool CMusicThumbLoader::FillLibraryArt(CFileItem &item)
         artname = artitem.prefix + "." + artitem.artType;
       }
 
+      // Pull out art for this specific disc e.g. "thumb2", skip art for other discs
+      if (tag.GetType() == MediaTypeAlbum && tag.GetDiscNumber() > 0 && tag.GetType() == artitem.mediaType)
+      {
+        // Find any trailing digits
+        size_t startnum = artname.find_last_not_of("0123456789");
+        std::string digits = artname.substr(startnum + 1);
+        int num = atoi(digits.c_str());
+        if (num > 0 && startnum < artname.size())
+        {
+          if (num == tag.GetDiscNumber())
+            discartmap.insert(std::make_pair(artname.substr(0, startnum+1), artitem.url));
+          continue;
+        }        
+      }
+
       artmap.insert(std::make_pair(artname, artitem.url));
 
       // Add fallback art for "thumb" and "fanart" art types only
@@ -319,6 +335,16 @@ bool CMusicThumbLoader::FillLibraryArt(CFileItem &item)
     if (!fanartfallback.empty())
       item.SetArtFallback("fanart", fanartfallback);
 
+    // Insert or replace album art with specifc disc art when we have some
+    for (auto discart : discartmap)
+    {
+      std::map<std::string, std::string>::iterator it = artmap.find(discart.first);
+      if (it != artmap.end())
+        it->second = discart.second;
+      else
+        artmap.insert(discart);
+    }
+
     item.AppendArt(artmap);
   }
```

I'll upload to my repo too if that is easier (I warned my Git use is a little rough'n'ready)

DaveTBlake commented:
OK, my art suggestion is in this repo https://github.com/DaveTBlake/xbmc/commits/AddBoxedsets4Art - dropped your art commit, added my own. Hopefully with my comments here you can see what this is about.  

Kodi will already (with the correct as.xml settings) pick up local art of all types from the CD folders and that CMusicThumbLoader::FillLibraryArt change is all that is needed to get it to show on the discs view. No idea if scraping disc specific art is a possility in the future, or get an art addon to fetch it.

We can talk more about getting multi-disc art from a single mixed disc folder (for now you could do that manually).

the-black-eagle commented:
Clever.  I wonder if we could use strings.po #427 (which is 'disc') to localize that and pick up any art that starts with that string ?  so 'disc.png' or 'discart.png' would work, regardless of any language used ?

DaveTBlake commented:
I would not bring strings.po into art types, users are confused enough (and some change language settings during use, imagine what thay would do).

Art types need to be in settings, something configurable not hard coded, somewhere all addons and skins can access, common settings for all (music) scrapers  and for automatic pick up of local art. At the moment that place is a whitelist in advancedsettings.xml, but that is not user friendly. It was just an initial step to get this functionality started in v18. For v19 I need to do more (you can help if you like).

I know users can set a variety of file names for thumbs, but once there is also a infinite list of possible art types having an infinite list of filenames for those types was all too much! Hence the simple map of filename to type. Add "thing" to as.xml, and locate thing.jpg appropriately (or use addon/scraper that fetches thing art), use a skin that shows "thing" art, and away you go. Core code changes never needed.  

The hope is that between skins and art addons the authors will manage to standardise on one name for round (disc) images - "discart" (is my suggestion) - but core does not impose that, it is up to the external community to sort it out. Or they can divide into garthering a varity, core doesn't care.

The legacy is "thumb" and "fanart", which for backwards compatibilty will probably always have more options, although internaly fanart will be handled less specifically than it currently is.

Sorry, hasty description before bed. I probably missed something or assumed something.

the-black-eagle committed:
Use DaveBlakes artwork changes

DaveTBlake commented:
Have updated my repo https://github.com/DaveTBlake/xbmc/tree/AddBoxedsets4Art with new art processing in Thumbloader specifically https://github.com/DaveTBlake/xbmc/commit/276920c9a6b2b998b31c6fd344a8927e3a84c457 replacing my previous attempt. See what you think, meanwhile I'm away a couple of days

the-black-eagle committed:
Make discart work when playing songs

DaveTBlake commented on xbmc/music/MusicThumbLoader.cpp:
```diff
@@ -322,6 +322,17 @@ bool CMusicThumbLoader::FillLibraryArt(CFileItem &item)
+          int number = atoi(artitem.artType.substr(7).c_str());
+          if (number == 0)
+            item.SetArtFallback("discart", artname);
+          else if (number > 0 && tag.GetDiscNumber() == number)
```
Hard coded art type again. What is needed muct work regardless of what the disc art gets called by scraper, skin or user, and even support multiple types of disc specific art (OK I can only think of a CD image, but who knows what furture things we could have??)

Also the fallback mechanism is historic, this only expected to be applied to thumbs/fanart. Yes it looks tempting approach, but can be achieved more obviously I think

DaveTBlake replied:
Take a look at my reworking, we possibly cross posted.

the-black-eagle replied:
Yeah, I will.  In regard to the hard-coding, I can see no other way with what there is at the moment to set a specific 'something%i', where %i may or not exist, into Player.Art(discart), which is what a skin is expecting.  In the art table, it is of type 'Album' and AS.xml has artistextraart and albumextraart.  If however, there was a type of 'Disc' and discextraart, then one could distinguish discart from anything else by the type.  Otherwise, I don't see how to separate out discart from thumb or cover or any other string that could have a numerical suffix and is of type Album without having a hard-coded string in there.  Thats before I've looked at what you've done though, so perhaps I'm talking without my book !!  

I didn't think that you would go for this anyway :)  

Just out of interest, is there any way of deleting art from the art table when something changes, without editing the db ?  In testing I ended up with discart and discart1 for a couple of albums and found that discart was always displayed (as it is found first) and the only way to change that was edit the db and drop that row......or at least, I think it was the only way.

the-black-eagle replied:
Ha, well looks like I was completely wrong then.  Just tested with 4 different skins with albums with multiple and single discs and all displayed correctly.  Now I need to go and pick your code apart to see how you've done it.

DaveTBlake replied:
Yes, pick that code apart, see if you can make sense of it or find more design omissions or mistakes :)

Try to think of art generically. When I say disc art I mean art associated with subpart of a release, an obvious example is the round CD image which was "cdart" for some addons/skins but we are suggesting standardising on "discart". However parts can also have inner sleave covers (taken from folder.jpg and held as "thumb") and could have other stuff in the future - wild example could be sleave back. Core is stuck with thumb (a defualt image of some kind), but does not want to hard code anything more.

Current skins may look for "discart", and scraper may fetch "discart" and local files called discart.png  will be picked up as "discart" if the right seeing is added to as.xml. The same core processing needs to work for "foobar" if that is what users/skins/scraper choose to implement as an art type.

Further we are treating a disc like a subset of an album with all the same properties plus a number and subtitle, hence same goes for art. Any art types an album can have ("thumb" + whatever is configured in as.xml) a disc can have, and are held as if for the album with disc number appended e.g. "thumb2", "foobar3" etc. So if a type of album art has a numberic suffix then it is disc (album part)  specific. What you reminded me was that the song needed its specific part of album art too.

> is there any way of deleting art from the art table when something changes,

From album info dialog, select art, or rather deselect that type. Of course all the possible art types (based on what types you have for some albums) get shown even when that specific item does not have that art as a way to allow users to select manually an use art type consistently. It is all still rather raw as a UI.

DaveTBlake commented:
Just a heads up that the next rebase against master there are a couple of merged PRs from me that will mean paying attention PR16626 and 16657. I have found it helps to know that the trip hazzard exists,  sorry about that, soon make a Git Ninja of you :-)

the-black-eagle commented:
Oooh, lovely !!!  I updated my master branch yesterday and then spent an (un)pleasant hour or so squashing, fixing, dropping and picking to get a nice fresh clean PR for you (#3).  Dropped the Estuary stuff along the way as I can do that later.  It also includes some code to allow a user to turn off disc navigation but not for boxsets. I'm now looking at how smartplaylists work, but need to have a serious look at filtering and get that working at the same time.
